File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.util.*;
import java.io.*;
class Main{
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int answer = 0;
        int[] wines = new int[n+1];
        int[] dp = new int[n+1];
        for(int i=0; i<n; i++) {
            wines[i+1] = Integer.parseInt(br.readLine());
        }
        dp[0] = 0;
        if(n == 1) {
            System.out.println(wines[1]);
            return ;
        }
        dp[1] = wines[1];
        dp[2] = wines[1] + wines[2];
        for(int i=3; i<=n; i++) {
            dp[i] = Math.max(Math.max(dp[i-2],  wines[i-1] + dp[i-3]) + wines[i], dp[i-1]) ;
        }
        System.out.println(dp[n]);
    }
}

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys
wine = sys.stdin.readline

n = int(input())
wines = [0]
dp = [0] * (n+1)
for _ in range(n):
    wines.append(int(wine()))

dp[1] = wines[1]
if n != 1 :
    dp[2] = wines[1] + wines[2]

    for i in range(3, len(wines)):
        dp[i] = max(max(dp[i-2], dp[i-3] + wines[i-1]) + wines[i], dp[i-1])

print(dp[-1])
